### ISSUE TITLE
Event onWithOptions doesn't allow selective behaviour on events

### ISSUE THREAD

**Issue by BarelyFunctional** [elm-lang/virtual-dom#18]
Sometimes its necessary to stopPropogation and preventDefault on certain events and not others. For instance it may be desired to restrict certain key-presses in a text-input field. Examples include:

Credit card entry, numbers and spaces
Phone numbers with international codes () and +
Hex editor, restrict to A-Z 0-5
and many more....
Currently Elm doesn't allow this behaviour because stopPropogation and preventDefault are specified at the time the event listener is registered, and before the event listener has received any events.

I suggest adding a new Options type, something like this:

`type alias EventOptions = { stopPropagation : Json.Decoder Bool , preventDefault : Json.Decoder Bool }`

and a companion method:

`onWithEventOptions`

This will allow stopPropogation and preventDefault flags to be based on the content of the event received.

**Comment by glenjamin** [comment]
I think another possible signature could be

`onAGoodName : String -> Json.Decoder (Options, msg) -> Html.Attribute msg`

Another scenario I came across recently: application hotkeys, you only want to `preventDefault` for certain keys.

**Comment by rtfeldman** [comment]
This seems like a sweet API! Since this is an unusual and advanced feature, I'd personally just replace the existing API with this one.

It wouldn't be much trouble to ca `onWithOptions` passing `{ stopPropagation = Decode.succeed True }` in the current case.

**Comment by lessless** [comment]
Hi guys! 

We get into situation where we have an Elm widget inside a form and which  should process `Enter key press` event in its inputs, perform custom logic *and* prevent form submission. 
The thing is that we can't select event on which we want to `preventDefault` and effectively blocking all input in those inputs.

**Comment by BarelyFunctional** [comment]
Hi Lessless, are you saying the proposed solution would address your issue?

**Comment by lessless** [comment]
I can't  make any judgement on the proposed way/solution, but 

>  allow stopPropogation and preventDefault flags to be based on the content of the event received.

is a super-essential, must-have, uber-needed feature.

**Comment by opsb** [comment]
One for the examples list:

I have a textarea and I want to trigger an update when the enter key is hit but prevent the newline from being added to the textarea.

**Comment by OvermindDL1** [comment]
As long as we can use the model passed in to the view too then that seems like a great API (and it looks like we can).  Whether I would cancel an event or not is quite often based on my state.

**Comment by anagrius** [comment]
We are hitting this problem too with an advanced textarea where we need to do preventing on a per event basis. 👍

**Comment by jackysee** [comment]
In my case, I need to trigger a msg by a keyboard shortcut `space`, so I need to preventDefault on the scrollable area to prevent the scrolling by browser.

In js, it's like
```
el.addEventListener('keyup', function(ev){
  if(ev.keyCode === 32){
    ev.preventDefault();
    //... to the stuff 
  }
});

**Comment by nicknovitski** [comment]
My use case is that I'm making random html elements act as slider inputs, and for the sake of accessibility, I want them to respond to all the keypresses listed in the WAI-ARIA "slider" role (the direction keys, pgup, pgdown, home and end).  While the elements have focus, the default of key events for those keys should be prevented, but any other keys should have their normal effect, or else accessibility is degraded rather than improved.

**Comment by Namek** [comment]
Making `stopPropagation` just a decoder receiver through `Options` doesn't cover all situations. Next usecase:

```
handleInput: function (event) {
      // only allow numeric keys
      if (event.keyCode < 48 || event.keyCode > 57) {
        event.preventDefault();
      }

      event.stopImmediatePropagation()
    },
```

registered to both `keypress` and `input` events for `<input type="text">.

IMO in the end we'll need access to `event` object and all it's properties and methods. Is there any hack to workaround to this? I'm new to Elm, can it be done through Ports somehow?

**Comment by aforemny** [comment]
Is this on the road-map for Elm 0.18? This would be the most important feature for me to have.

Basically, preventing the default for input elements is not possible, because as of now it will prevent Tab navigation. Preventing browser actions conditionally is all over the JS world and a must-have.

**Comment by lessless** [comment]
Well, if the language is perceived as a technology that will be used 20
years after, then this feature can wait couple of years ;)))

On Nov 14, 2016 8:24 PM, "Alexander Foremny" <notifications@github.com>
wrote:

Is this on the road-map for Elm 0.18? This would be the most important
feature for me to have.

Basically, preventing the default for input elements is not possible,
because as of now it will prevent Tab navigation. Preventing browser
actions conditionally is all over the JS world and a must-have.

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/elm-lang/virtual-dom/issues/18#issuecomment-260333239>,
or mute the thread

**Comment by aforemny** [comment]
The current API for events is debilitating. OP's original proposal is a sensible extension of the current API that seems to solve all but one mentioned use-case here (Namek's). However, that one is about the separate issue of extending event handling options by `stopImmediatePropagation`.

As you seem like a strong supporter of this issue, @lessless, based on your original comment above, I understand your comment as an expression of frustration. I would like to state, however, that because of this issue, it is very cumbersome to work any further on elm-mdl, and it seriously prevents me from even thinking about introducing Elm in work-space.

**Comment by Namek** [comment]
There are non-publish components I've written which I would need to have still made totally in JavaScript. Just because of this single thing I can't port them to Elm. Communication through ports doesn't seem to solve everything, especially if it introduces umaintainable spaghetti (or lasagne?).

There's next thing I didn't mention previously but had in mind. Elm seems to assume today's standards like HTML5 or ES6. And I think it's good BUT sometimes we need to support little order browser where events don't have stopPropagation() but have instead `cancelBubble = true` -> https://developer.mozilla.org/en-US/docs/Web/API/UIEvent/cancelBubble Sure, it may be deprecated or non-standard but it's there.

I understand that Elm wants to avoid undefined things. To cover this situation, maybe Ports should be somewhat extended. Perhaps, some additional function could exist that would send not just a message but also (in this particular case) a reference to the event object. So, in the JavaScript side we could call anything we want on given event object. The problem is, we need to do this in this very fram, I mean, it can't be passed to JS event loop, it has to be called NOW. Sure, it IS possible it would introduce runtime exception but some Native part of Elm in this could catch this and continue.

To sum up, this conditional `stopPropagation()` thing is a matter of "throw and forget". Thinking about it that way shouldn't introduce runtime errors triggered by Elm code, I think.

**Comment by chancyk** [comment]
I've run into this issue when attempting to allow tab navigation of a popup menu raised by a search box, so it's either block all events handled by the `input` element using `onWithOptions` or have the browser steal focus after the first press of tab.

It would also be nice to have a similar granularity of control on the more global `Keyboard.presses` and its sibling subscriptions.

**Comment by jackysee** [comment]
Another case here:  an autocomplete field needs to support 'up' and 'down' to navigate through result, while keeping the cursor position in the input field. Currently pressing 'up' would make the cursor goes to the leftmost position. You need a conditional `stopPropgation()` here to allow typing of character but 'up' and 'down'

**Comment by mreriklott** [comment]
Forms - I have several forms that I'd like to submit normally, but I'd like elm to prevent submission if the form model is invalid....

```
onInvalidSubmit : Model -> Msg -> Attribute Msg
onInvalidSubmit model msg =
    let
        options =
            if isValid model then
                { stopPropagation = False, preventDefault = False }
            else
                { stopPropagation = True, preventDefault = True }
    in
        onWithOptions "submit" options (Json.succeed msg)
```

This is simple to express in elm using the "onWithOptions" function, but it doesn't seem to work. The submit event is always prevented.

**Comment by smurfix** [comment]
OK, we've seen a lot of use cases here. I could contribute my own but that wouldn't be helpful.

Any progress on actually solving this problem? Other than using "real" JS to handle the event in question … always a possible workaround, but not one I'm particularly fond of.

**Comment by lessless** [comment]
I've heard that it will be implemented natively in WebAssembly release

On Sun, Jan 15, 2017 at 10:06 PM, Matthias Urlichs <notifications@github.com

**Comment by inactivist** [comment]
@lessless -- can you provide additional info on this?  I'm not sure what that means.

**Comment by lessless** [comment]
@inactivist that's a sarcastic comment, it means we shouldn't hope to see this feature anytime soon. 
"Compilation into WebAssembly" is used to crude how long it might take because, obviously, WebAssembly is not even on the roadmap either.

**Comment by sentience** [comment]
Here’s the work-around we’ve been using to `preventDefault` on only certain `"keydown"` events.

First, on the intended target element, do a “normal” `on "keydown"` that identifies the desired keystrokes and responds to them, ignoring keystrokes that are not of interest to us. This listener allows all `"keydown"` events to continue propagating with their default action enabled.

```elm
input
    [ type "text"
    , on "keydown" (Decode.map KeyDown keyCode)
    ]
    []
```

Then, wrap the target element in a parent element with its own `"keydown"` event listener, this time using `onWithOptions` to prevent the default action for all _successfully decoded_ events. The trick, then (with thanks to @opsb for [this tip](https://github.com/elm-lang/html/issues/66#issuecomment-238022618)) is to only decode successfully the keycodes for which we wish to prevent the default action:

```elm
div
    [ preventDefaultUpDown
    ]
    [ input
        [ type "text"
        , on "keydown" (Decode.map KeyDown keyCode)
        ]
        []
    ]

⋮

preventDefaultUpDown : Html.Attribute (Msg record)
preventDefaultUpDown =
    let
        options =
            { Html.Events.defaultOptions | preventDefault = True }

        filterKey code =
            if code == KeyCodes.upArrow || code == KeyCodes.downArrow then
                Ok code
            else
                Err "ignored input"

        decoder =
            keyCode
                |> Decode.andThen (filterKey >> DecodeExtra.fromResult)
                |> Decode.map (always NoOp)
    in
        onWithOptions "keydown" options decoder
```

**Comment by simonh1000** [comment]
My use case for this feature is with contenteditable divs. I want to trap an <enter> and instead of entering a `char 13`, use this as a trigger to blur the selected node and save the `evt.target.text` value back to the model

**Comment by MethodGrab** [comment]
The work around I ended up using for this problem was to add an inline JavaScript handler like so:
Runnable example: https://ellie-app.com/3t42RzdkDQRa1/7

```elm
module Main exposing (..)

import Html exposing (..)
import Html.Attributes exposing (..)
import Html.Events exposing (..)


main : Program Never Model Msg
main =
    beginnerProgram { model = init, view = view, update = update }


type alias Model =
    { counter : Int
    }


init : Model
init =
    { counter = 0
    }


type Msg
    = NoOp
    | Increment


update msg model =
    case msg of
        NoOp ->
            model

        Increment ->
            { model | counter = model.counter + 1 }


view : Model -> Html Msg
view model =
    let
        inlineOnClick =
            -- the inline onclick *attribute* needs to come before the elm onclick handler
            -- you can inline the function directly or refer to a global function by name eg:
            -- [ attribute "onclick" "preventDefaultIfCtrl(event)" ]
            [ attribute "onclick" "if (event.ctrlKey) { event.preventDefault(); event.stopImmediatePropagation(); }" ]

        buttonAttrs =
            List.concat [ inlineOnClick, [ onClick Increment ] ]
    in
        div []
            [ p [] [ text "Hold down the `CTRL` key when clicking the increment button to prevent the default behavior" ]
            , button buttonAttrs [ text "Increment" ]
            , div [ class "count" ]
                [ text ("Count: " ++ toString model.counter)
                ]
            ]
```

**Comment by evancz** [comment]
The upcoming API in 0.19 has [`onBubble`](https://github.com/elm-lang/virtual-dom/blob/dev/src/VirtualDom.elm#L230-L244) and [`onCapture`](https://github.com/elm-lang/virtual-dom/blob/dev/src/VirtualDom.elm#L247-L255) that allow more flexibility.

In particular, they both take a [`Handler`](https://github.com/elm-lang/virtual-dom/blob/dev/src/VirtualDom.elm#L258-L279) that gives you the ability to do `stopPropagation` and `preventDefault` conditionally. This design also means we can create "passive" listeners automatically, making scrolling smoother in touch devices.

If you think there are cases that the new API will not cover, please share them here. If others validate that the new API would not cover them, please open a new issue outlining the scenario before 0.19 comes out!

**Comment by sh4r3m4n** [comment]
Hi! I'm using ports to do something like this and waiting for 0.19 so maybe I can do this natively without ports. I noticed that there is a way to conditionally call `stopPropagation()`, but using [`stopImmediatePropagation()`](https://developer.mozilla.org/en-US/docs/Web/API/Event/stopImmediatePropagation) instead stills not possible (as @Namek pointed out a time ago).

I'm currently using that function in an event handler of a content script of a webextension, to prevent listeners attached to the **same** element from executing (`stopPropagation` wouldn't do that).
Actually, I am doing this in a global (document) event handler so I would use this from elm-lang/dom instead of from elm-lang/virtual-dom, but I comment it here because I understand that both packages should have the same API to define event handler behaviors.

If I had to propose an improvement to support this case, I think adding a `stopImmediatePropagation` field in the Custom Handler decoded type would work:
```elm
type Handler msg
-- ....
| Custom (Decode.Decoder { message : msg, stopPropagation : Bool, stopImmediatePropagation: Bool, preventDefault : Bool })
```
Another approach would be changing the Bool of MayStopPropagation to a custom trinary union type:
```elm
type Handler msg
  = Normal (Decode.Decoder msg)
  | MayStopPropagation (Decode.Decoder (msg, StopPropagationAction))
-- ....
type StopPropagationAction
    = NoStopPropagation
    | StopChildPropagation
    | StopImmediatePropagation
```
